### ISSUE TITLE
Model validation scheme

### ISSUE THREAD

**Issue by kevinykuo** [kevinykuo/MLTMS#1]
Discussion for determining model validation scheme.

**Comment by kevinykuo** [comment]
There have been a few email exchanges within one of the subgroups on training data selection. I'll provide some general comments here and encourage others to do the same.

1. At the end of a calendar year, the reserving actuary is only concerned with accidents occurred up to that year; you can't put on the books the losses that will happen in the future (practicing actuaries please comment if this is correct; there may also be geographical difference.) If we're standing at the end of, for example, AY 15, the changes in the environment (legislative, inflation, etc.) that will affect AY 16 and on are irrelevant to us.

  * Let's say at the end of AY 15 there was a legislative change. At the end of year 16 the model will have only seen one diagonal in the new environment. It may perform poorly in forecasting development for AY 16. We may attempt to ameliorate this by training on data where previous environment changes have happened. If we are unsuccessful we would simply concede this as a shortcoming of the model.

2. The models (or modeling pipelines) that we come up with need to be able to dynamically evaluate available data and forecast claims development at arbitrary valuation dates. In other words, if we decide that the valuation interval is yearly, the models should perform the forecast at the end of each accident year where we have enough historical data to evaluate the performance of the forecast.

3. We would still need to decide on a point in time where information available after that point is completely withheld from the model building process. A temporal divide of information is crucial. During production, we won't have visibility into what e.g. legislative changes will occur in the future, so we cannot incorporate that in the modeling process.

4. The train/validation split decisions for *hyperparameter tuning* should be left to the subgroups constructing the models. The model validation subgroup should only be concerned with how the models will be *evaluated*.

**Comment by LorenzoInvernizzi** [comment]
Hi All,

I would like to point out some personal thoughts regarding the evaluation methods. I think this is also a ML choice, hence fell free to participate in the discussion (@Hong, @Zack Martin).

I assume that we chose the tree based models (Random Forest and/or gradient boosted trees). However, the ideas remain the same as for NN or others ML models.

What type of validation are we going to use in order to tune the hyper parameters? Some possible are:
-    Simple validation, ie 1 fold cross validation
-    K fold cross validation 
After computing the optimal set of hyper parameters we could also use the Super Learner approach of Van Der Laan (suggested by @Salma Jamal), instead of fitting the model with the optimal parameters on the entire train set (even the validation set I think).

From my experience with the H2o package, the k-fold CV can be easily done to tune the number of trees in the model. It does not handle the Super Learner approach. However I found an extestion of the h2o package that handles this approach (still need to study it). See for reference: http://docs.h2o.ai/h2o-tutorials/latest-stable/tutorials/ensembles-stacking/index.html

The main point is that depending on how we divide the dataset we have different scenarios:
- individual claims approach: after rearranging the datasets in columns I can do whatever random split I want, keeping the same portion of data on each accidental years and also the development year.
- sub triangles approach (@Greg Taylor): from my point of view it is difficult to identify a validation set. We could consider for example the next or previous 2 years after the splitting on time of the dataset as a validation set. Here k-fold CV does not make sense I think. What do you think?

I pointed out some personal thoughts, fill free to comment and add your experience on this

**Comment by zack** [comment]
@ZJMartin ^^

**Comment by kevinykuo** [comment]
@LorenzoInvernizzi h2o supports stacked ensemble natively via `h2o.stackedEnsemble()`, see http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/stacked-ensembles.html.

From the *model validation* (not hyperparameter tuning) perspective there should be one (or more) measure(s) that's used evaluate all methods/models consistently. There should not be a claims level measure (since you can't do claims-level predictions with chain ladder) and any such measure wouldn't take into account IBNR.

The deliverables of the model validation should include something along the lines of a specification of what the modeling subgroups should produce as predictions (e.g. aggregate unpaid claims by AY as of 10 consecutive year-end valuation dates), and a function that takes in some predictions and spits out a performance metric.

**Comment by hongshen1010** [comment]
@LorenzoInvernizzi, @kevinykuo. Thanks for your thoughts on model validation scheme. 

If our modeling target is the incremental transaction of individual claims, then we may do a calendar year separation for the training/validation/test. We can have multiple training/validation sets with such an approach.

For example, we can have calendar year 1981 - 1990 as the training  and calendar year 1991 - 1995 as the validation. Then we can move 5 years forward to have 1986 - 1995 as training and calendar year 1996 - 2000 as validation. Another sets for training and validation can be just another shift of 5 years. And we can leave calendar year 2011 - 2017 as the test dataset. 

Such a calendar year split will be consistent with reserving need as our prediction for the current triangle is for future calendar years. Also the multiple sets for training/validation can help to measure the model stability over time and prevent overfitting.

Just my two cents on this topic.

**Comment by MunirHiabu** [comment]
Thanks everyone for your inputs. 

I would like to summarize a few points:

1) As pointed out by others, We should not mix the dicussion of model
    validation and validation within a training set for parameter tuning.
  The latter is soemthing that should be discussed within the machine learning group.
   But since it was mentioned just a few remark from me abot the machine learning training (which you can actually also find hinted in the word document).
  
   Choosing the right paramter in a machine learning method is a bias-variance trade-off decision.
   The 'best' parameter will heavily depend on how the model will later be evaluated.
   This means, if we just compare the full reserve, you will need other parameters than in the case where we compare 1yr or 5yr chashflows.
   The approach Hong mentioned makes for instance sense if we are interested in 5yr cashflow forecasts.
   Finally, cross-validation or other random splits of the training set will perform poorly since then the method is trained for a wrong task.






2) A random 80%-20% split is not suitable for our resreving problem.
    Also a split via accident years is not suitable.
    As a forecasting problem, we need to move along calendar time.
    Please reread the uplaoded Word document for more details.


3)   As an example if we decide to have 1980-2004 as training and 2005-1016 as test, then the traing set should consist of all
     data that was available to the company at the end of 2004. As a consequence as pointed out by Salma, for instance in this example,
    we will have to ignore all claims with accident date after 2004.
